src/ibek/ioc.py

```diff
@@ -54,64 +55,75 @@ class Entity(BaseSettings):
-        ids = {a.name for a in self.__definition__.args if isinstance(a, IdArg)}
+        if self._model.pre_defines:
+            for name, define in self._model.pre_defines.items():
+                self._process_field(name, define.value, define.type)
 
-        entity_dict = self.model_dump()
-        for arg, value in entity_dict.items():
-            model_field = self.model_fields[arg]
+        if self._model.parameters:
+            for name, parameter in self._model.parameters.items():
+                self._process_field(name, getattr(self, name), parameter.type)
 
-            if isinstance(value, str):
-                # Jinja expansion of any of the Entity's string args/values
-                value = UTILS.render(entity_dict, value)
-                # this is a cheesy test - any better ideas please let me know
-                if "Union" in str(model_field.annotation):
-                    # Args that were non strings and have been rendered by Jinja
-                    # must be coerced back into their original type
-                    try:
-                        # The following replace are to make the string json compatible
-                        # (maybe we should python decode instead of json.loads)
-                        value = value.replace("'", '"')
-                        value = value.replace("True", "true")
-                        value = value.replace("False", "false")
-                        value = json.loads(value)
-                        # likewise for bools
-                    except:
-                        print(
-                            f"ERROR: fail to decode {value} as a {model_field.annotation}"
-                        )
-                        raise
-                setattr(self, arg, value)
-                # update the entity_dict with the rendered value
-                entity_dict[arg] = value
-
-            if model_field.annotation == object:
-                # look up the actual object by it's id
-                if isinstance(value, str):
-                    value = get_entity_by_id(value)
-                    setattr(self, arg, value)
-                # update the entity_dict with looked up object
-                entity_dict[arg] = value
-
-            if arg in ids:
-                # add this entity to the global id index
-                if value in id_to_entity:
-                    raise ValueError(f"Duplicate id {value} in {list(id_to_entity)}")
-                id_to_entity[value] = self
+        if self._model.post_defines:
+            for name, define in self._model.post_defines.items():
+                self._process_field(name, define.value, define.type)
 
         return self
 
     def __str__(self):
         # if this entity has an id then its string representation is the value of id
-        id_name = self.__definition__._get_id_arg()
+        id_name = self._model._get_id_arg()
         return getattr(self, id_name) if id_name else super().__str__()
```

Comment:
What does `super().__str__()` end up being? When the Entity doesn't have an id?

Reply:
The intention was that it would call the str() for int or float but that's wrong as the base class is supposed to be Entity.

I think the check is redundant because you cant refer to another entity unless it has an id. Therefore this should raise and error - but will never be hit

Reply:
I have changed this to a raise.